nnvm/python/nnvm/testing/check_computation.py

```diff
@@ -0,0 +1,451 @@
+            shape[x_name] = x[1]
+            x = x[0]
+
+        grad_input_vars_real.append(input_dict[x] if isinstance(x, str) else x)
+
+    grad_input_vars = grad_input_vars_real
+
+    # Infer the output shape and dtype by creating a graph and running passes
+
+    forward_graph = nnvm.graph.create(symbol)
+
+    if dtype is not None:
+        nnvm.compiler.graph_attr.set_dtype_inputs(forward_graph, dtype)
+
+    if shape is not None:
+        nnvm.compiler.graph_attr.set_shape_inputs(forward_graph, shape)
+
+    forward_graph = forward_graph.apply('InferShape').apply('InferType')
+    shapes = forward_graph.json_attr('shape')
+    dtypes = forward_graph.json_attr('dtype')
+    out_shape = shapes[forward_graph.index.output_entries[0][0]]
+    out_dtype = dtypes[forward_graph.index.output_entries[0][0]]
+
+    backward_graph = None
+
+    # If we want gradients, we have to recreate the graph, but now with gradient computations
+    # Note that here we need out_shape for defining the shape of head grads, so we have to
+    # create the graph twice
+    if np_backward is not None or numerical_grads:
+        try:
+            head_grads_symbol = nnvm.symbol.Variable("head_grads", shape=out_shape, dtype=out_dtype)
+            grad_symbols = graph_util.gradients([symbol], grad_input_vars,
+                                                grad_ys=head_grads_symbol)
+            # Sometimes grads do not depend on head_grads, so head_grads does not appear
+            # in the variable list; adding it manually prevents this, making things a bit easier
+            backward_graph = \
+                nnvm.graph.create(nnvm.symbol.Group([symbol] + grad_symbols + [head_grads_symbol]))
+
+            if dtype is not None:
+                nnvm.compiler.graph_attr.set_dtype_inputs(backward_graph, dtype)
+
+            if shape is not None:
+                nnvm.compiler.graph_attr.set_shape_inputs(backward_graph, shape)
+
+            backward_graph = backward_graph.apply('InferShape').apply('InferType')
+            shapes = backward_graph.json_attr('shape')
+            dtypes = backward_graph.json_attr('dtype')
+        except nnvm._base.NNVMError as err:
+            if np_backward is None and numerical_grads == "if_possible":
+                logging.warning("Won't check gradients because: %s", str(err).split('\n', 1)[0])
+                numerical_grads = False
+                backward_graph = None
+            else:
+                raise
+
+    main_graph = backward_graph if backward_graph is not None else forward_graph
+
+    if dump_graph:
+        print()
+        print(main_graph.ir(join_node_attrs=['shape', 'dtype']))
+
+    # Generate random data for inputs (including head_grads)
+
+    np_inputs = {}
+
+    for x in main_graph.symbol.list_input_variables():
+        x_name = x.attr('name')
+
+        low = -1.0
+        high = 1.0
+        if in_range is not None:
+            if isinstance(in_range, dict):
```

Comment:
If it is just for random ndarray generation, it might be better to just set in_range as (low, high) directly? Will it make much difference for different in_range for inputs? This can make the logic much simpler.

Or we can simply define it as dict from str to num array.